src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs

```diff
@@ -44,6 +205,469 @@ private void BreakPointWithDebuggerFuncEval()
+                /// to efficiently fetch that property from a .NET object (See Fetch method).  
+                /// It hides some slightly complex generic code.  
+                /// </summary>
+                class PropertyFetch
+                {
+                    /// <summary>
+                    /// Create a property fetcher from a .NET Reflection PropertyInfo class that
+                    /// represents a property of a paritcular type.  
+                    /// </summary>
+                    public static PropertyFetch FetcherForProperty(PropertyInfo propertyInfo)
+                    {
+                        if (propertyInfo == null)
+                            return new PropertyFetch();     // returns null on any fetch.
+
+                        var typedPropertyFetcher = typeof(TypedFetchProperty<,>);
+                        var instantiatedTypedPropertyFetcher = typedPropertyFetcher.GetTypeInfo().MakeGenericType(
+                            propertyInfo.DeclaringType, propertyInfo.PropertyType);
+                        return (PropertyFetch)Activator.CreateInstance(instantiatedTypedPropertyFetcher, propertyInfo);
+                    }
+
+                    /// <summary>
+                    /// Given an object, fetch the property that this propertyFech represents. 
+                    /// </summary>
+                    public virtual object Fetch(object obj) { return null; }
+
+                    #region private 
+
+                    class TypedFetchProperty<Tobject, Tproperty> : PropertyFetch
+                    {
+                        public TypedFetchProperty(PropertyInfo property)
+                        {
+                            _propertyFetch = (Func<Tobject, Tproperty>)property.GetMethod.CreateDelegate(typeof(Func<Tobject, Tproperty>));
+                        }
+                        public override object Fetch(object obj)
+                        {
+                            return _propertyFetch((Tobject)obj);
+                        }
+                        private Func<Tobject, Tproperty> _propertyFetch;
+                    }
+                    #endregion
+                }
+
+                string _propertyName;
+                Type _expectedType;
+                PropertyFetch _fetchForExpectedType;
+                #endregion
+            }
+
+            string _outputName;
+            PropertySpec _fetches;
+            #endregion
+        }
+
+        /// <summary>
+        /// CallbackObserver is a adapter class that creats an observer (which you can pass
+        /// to IObservable.Subscribe), and calls the given callback every time the 'next' 
+        /// operation on the IObserver happens. 
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        internal class CallbackObserver<T> : IObserver<T>
+        {
+            public CallbackObserver(Action<T> callback) { _callback = callback; }
+
+            #region private 
```

Comment:
Yes, that would make it purer.  It does not work on abstract types or overrides (e.g. ToString).  It also has the side effect of making it harder to call the implemented method (you have to cast it to the interface before making the call).   

But mostly I was just trying to do something simple (hide methods the client does not care about), and don't really want to perturb the code just to make it pure.   I could name the region 'non-client' which more accurately describes the partitioning, but without explaination, it is probably no better than 'private'.     

For what it is worth, if push came to shove I would rather simply move the public things out of the private region, but it often does clutter the client interface (since we put thing in the private region because they are more implementation than client contract).